### ISSUE TITLE
`invalid-repr-returned` false positive with match statement

### ISSUE THREAD

**Issue by CobaltCause** [PyCQA/pylint#7645]
### Bug description

It looks like pylint isn't checking exhaustiveness before triggering this lint.

`pylint_e0306_false_positive.py`:
```python
# pylint: disable=missing-docstring,too-few-public-methods

from typing import Optional


class FalsePositive:
    def __init__(self) -> None:
        self.opt_field: Optional[int] = 5

    def __repr__(self) -> str:
        match self.opt_field:
            case None:
                return "None"
            case some_value:
                return f"Some({repr(some_value)}"
```

Likely related: https://github.com/PyCQA/pylint/issues/7470#issuecomment-1266292154

### Command used

```shell
pylint pylint_e0306_false_positive.py
```

### Pylint output

```shell
************* Module pylint_e0306_false_positive
pylint_e0306_false_positive.py:10:4: E0306: __repr__ does not return str (invalid-repr-returned)
```

### Expected behavior

No error

### Pylint version

```shell
pylint 2.14.5
astroid 2.11.7
Python 3.10.7 (main, Sep  5 2022, 13:12:31) [GCC 11.3.0]
```

### OS / Environment

NixOS 22.05

**Comment by Pierre-Sassoulas** [comment]
If nothing is matched this function would return None. This is hard to understand for pylint but also for human. Even for human you can make a mistake and not realize something can end up unmatched and get an error. I would suggest using a default.

**Comment by Pierre-Sassoulas** [comment]
> I don't believe it's correct to require a default value for an unreachable branch, at best that's just adding dead code for no reason.

The reason would be that you know within 10ms when reading the code that it can't return ``None`` instead of having to think about each possible case or blindly trust the tooling. I don't think a codebase should rely too much on trust in their tooling (trust but verify instead ?). 

That being said, this is a false positive that should be fixed, but this rely on control flow (#4795 ?). Control flow understanding is one of the weak point of pylint.

**Comment by CobaltCause** [comment]
I "know within 10ms when reading the code that it can't return `None`" because I don't see any errors in my editor informing me that my `match` was nonexhaustive. I trust this result, but can easily verify it by scanning the `case`s and comparing to the type in the `match`. Generally, the failure mode for tools around exhaustiveness checking is false negatives rather than false positives, so I think it's pretty safe to rely on the tool here. I don't think this line of reasoning holds any water.

---

> On the other hand, maybe this lint shouldn't be the responsibility of pylint, but that of a typechecker, since this is fundamentally about the type system and not code style.

I'd like to take a stronger stance on this and say the proper solution to this issue is to simply remove this lint and let a typechecker handle it instead.

**Comment by Pierre-Sassoulas** [comment]
We have a lot of legacy checkers that half-ass checks that should be done by a type-checker in a very particular scenario. On the top of my head this is one of the worst offender : https://pylint.pycqa.org/en/latest/user_guide/messages/error/invalid-envvar-value.html. This is a lot of cleanup and/or documentation to do if it come to it.

**Comment by udifuchs** [comment]
Another example of this issue is with enum:
```
import enum

class Foo(enum.Enum):
    ONE = 1
    TWO = 2

    def __repr__(self) -> str:
        match self:
            case Foo.ONE:
                return "One"
            case Foo.TWO:
                return "Two"
```
The point is that I do not want to have a default `case _:` pattern because I want my linter (pylint or mypy) to warn me if there is an enum value that is missing a case pattern.

**Comment by jacobtylerwalls** [comment]
Verified OP and https://github.com/PyCQA/pylint/issues/7645#issuecomment-1292840162 are fixed with PyCQA/astroid#2042, so closing as a duplicate of #5288, which we'll close with a regression test when updating pylint to use the next astroid release.
